guides/source/active_model_basics.md

```diff
@@ -44,76 +67,331 @@ class EmailContact
+object as follows:
+
+```html+erb
+<%= form_with model: EmailContact.new do |form| %>
+  <%= form.text_field :name %>
+<% end %>
+
+# =>
+<form action="/email_contacts" method="post" data-remote="true">
+  <input type="text" name="email_contact[name]">
+</form>
+```
+
+`render` can be used to render a partial with the object as a local variable:
+
+```html+erb
+<%= render partial: "email_contact", email_contact: EmailContact.new(name: "David", email: "david@example.com", message: "Hello World") %>
+```
+
+### Attributes
 
-- model name introspection
-- conversions
-- translations
-- validations
+`ActiveModel::Attributes` allows you to define data types, set default values,
+and handle casting and serialization on plain Ruby objects. This can be useful
+for form data which will produce Active Record-like conversion for things like
+dates and booleans on regular objects.
 
-It also gives you the ability to initialize an object with a hash of attributes,
-much like any Active Record object.
+To use Attributes, include the module in your model class and define your
+attributes using the `attribute` macro. It accepts a name, a cast type, a
+default value, and any other options supported by the attribute type.
+
+```ruby
+class Person
+  include ActiveModel::Attributes
+
+  attribute :name, :string
+  attribute :date_of_birth, :date
+  attribute :active, :boolean, default: true
+end
+```
 
 ```irb
-irb> email_contact = EmailContact.new(name: 'David', email: 'david@example.com', message: 'Hello World')
-irb> email_contact.name
-=> "David"
-irb> email_contact.email
-=> "david@example.com"
-irb> email_contact.valid?
+irb> person = Person.new
+
+irb> person.name = "Jane"
+irb> person.name
+=> "Jane"
+
+# casts the string to a date set by the attribute
+irb> person.date_of_birth = "2020-01-01"
+irb> person.date_of_birth
+=> Wed, 01 Jan 2020
+irb> person.date_of_birth.class
+=> Date
+
+# uses the default value set by the attribute
+irb> person.active
 => true
-irb> email_contact.persisted?
+
+# casts the string to a boolean set by the attribute
+irb> person.active = "0"
+irb> person.active
 => false
 ```
 
-Any class that includes `ActiveModel::API` can be used with `form_with`,
-`render` and any other Action View helper methods, just like Active Record
-objects.
+Some additional methods described below are available when using
+`ActiveModel::Attributes`.
+
+#### Method: `attribute_names`
+
+The `attribute_names` method returns an array of attribute names.
+
+```irb
+irb> Person.attribute_names
+=> ["name", "date_of_birth", "active"]
+```
+
+#### Method: `attributes`
+
+The `attributes` method returns a hash of all the attributes with their names as
+keys and the values of the attributes as values.
+
+```irb
+irb> person = Person.new
+irb> person.name = "John"
+irb> person.date_of_birth = "1998-01-01"
+irb> person.active = false
+
+irb> person.attributes
+=> {"name"=>"John", "date_of_birth"=>Thu, 01 Jan 1998, "active"=>false}
```

Comment:
Maybe we could use the `person` we already have? To minimize the amount of code a reader has to process. Just a suggestion.

```suggestion
>> person.attributes
=> {"name"=>"Jane", "date_of_birth"=>Wed, 01 Jan 2020, "active"=>true}
```